src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/AbstractBitbucketClient.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2020 Marvin Wichmann
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+package com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.client;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.UnifyConfiguration;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.BitbucketServerPullRequestDecorator;
+import com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.client.model.server.ErrorResponse;
+import okhttp3.Response;
+
+import java.io.IOException;
+
+abstract class AbstractBitbucketClient {
+    private UnifyConfiguration configuration;
+
+    ObjectMapper getObjectMapper() {
+        return new ObjectMapper()
+                .setSerializationInclusion(JsonInclude.Include.NON_NULL)
+                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+    }
+
+    String baseUrl() {
+        return configuration.getRequiredProperty(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_URL);
+    }
+
+    String getToken() {
+        return configuration.getRequiredProperty(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_TOKEN);
+    }
+
+    void validate(Response response) throws IOException {
+        if (!response.isSuccessful()) {
+            ErrorResponse errors = null;
+            if (response.body() != null) {
+                errors = getObjectMapper().reader().forType(ErrorResponse.class)
+                        .readValue(response.body().string());
+            }
+            throw new BitbucketException(response.code(), errors);
+        }
+    }
+
+    public void setConfiguration(UnifyConfiguration configuration) {
+        this.configuration = configuration;
```

Comment:
This is likely to cause an issue: the instances created by the Pico container are singletons so you're risking a race condition by modifying the internal state of a decorator whilst it's potentially in use in another thread. You'll need to ensure your approach is stateless by either converting this to a generator/factory pattern, or using injected instances that are state aware (e.g. creating a `UnifyConfiguration` when required based on using an injected `Configuration` and `ScannerContext` on-demand)